// hardhat-ethernal/dist/Ethernal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ethernal = void 0;
const ethers_1 = require("ethers");
const message_trace_1 = require("hardhat/internal/hardhat-network/stack-traces/message-trace");
const api_1 = require("./api");
const ETHERNAL_API_ROOT = process.env.ETHERNAL_API_ROOT || 'https://api.tryethernal.com';
const ETHERNAL_WEBAPP_ROOT = process.env.ETHERNAL_WEBAPP_ROOT || 'https://app.tryethernal.com';
const logger = (message) => {
    console.log(`[Ethernal] `, message);
};
const handleError = (baseMessage, error, verbose) => {
    var _a;
    try {
        const errorMessage = ((_a = error.response) === null || _a === void 0 ? void 0 : _a.data) || (error === null || error === void 0 ? void 0 : error.message) || `Can't find an error message. Try in verbose mode.`;
        logger(`${baseMessage}: ${errorMessage}`);
        if (verbose)
            console.log(error);
    }
    catch (_error) {
        logger(_error.message);
        if (verbose)
            console.log(_error);
    }
};
class Ethernal {
    constructor(hre) {
        this.env = hre;
        this.api = new api_1.Api(ETHERNAL_API_ROOT, ETHERNAL_WEBAPP_ROOT);
        this.syncNextBlock = false;
        this.traces = [];
        this.verbose = hre.config.ethernal.verbose || false;
        this.provider = hre.ethers.provider;
        if (this.verbose) {
            console.log('### Verbose mode activated - Showing Ethernal plugin config ###');
            console.log(this.env.config.ethernal);
            console.log('### End of config ###');
        }
    }
    async startListening() {
        const envSet = await this.setLocalEnvironment();
        if (!envSet) {
            return;
        }
        this.provider.on('block', (blockNumber, error) => {
            if (!!this.env.config.ethernal.skipFirstBlock && !this.syncNextBlock) {
                logger(`Skipping block ${blockNumber}`);
                this.syncNextBlock = true;
            }
            else {
                this.onData(blockNumber, error);
            }
        });
    }
    async push(targetContract) {
        if (this.env.config.ethernal.disabled) {
            return;
        }
        const envSet = await this.setLocalEnvironment();
        if (!envSet) {
            return;
        }
        this.targetContract = targetContract;
        if (!this.targetContract.name || !this.targetContract.address) {
            return logger('Contract name and address are mandatory');
        }
        const contract = await this.getFormattedArtifact(targetContract);
        if (!contract) {
            return;
        }
        try {
            await this.api.syncContractData(contract.name, contract.address, contract.abi, targetContract.workspace);
        }
        catch (error) {
            handleError('Error syncing contract data', error, this.verbose);
        }
        if (this.env.config.ethernal.uploadAst) {
            logger('Uploading ASTs, this might take a while depending on the size of your contracts.');
            try {
                await this.api.syncContractAst(contract.address, {
                    artifact: contract.artifact,
                    dependencies: contract.dependencies
                }, targetContract.workspace);
            }
            catch (error) {
                handleError(`Couldn't sync dependencies`, error, this.verbose);
            }
        }
        const dependencies = Object.entries(contract.dependencies).map(art => art[0]);
        const dependenciesString = dependencies.length ? ` Dependencies: ${dependencies.join(', ')}.` : '';
        const workspaceString = ` Workspace: ${targetContract.workspace || this.api.currentWorkspaceName}.`;
        logger(`Updated artifacts for contract ${contract.name} (${contract.address}).${dependenciesString}${workspaceString}`);
    }
    async traceHandler(trace, isMessageTraceFromACall) {
        if (this.env.config.ethernal.disabled)
            return;
        if (this.env.config.ethernal.disableTrace)
            return;
        if (isMessageTraceFromACall)
            return;
        await this.setLocalEnvironment();
        const envSet = await this.setLocalEnvironment();
        if (!envSet)
            return;
        const parsedTrace = [];
        let stepper = async (step) => {
            if ((0, message_trace_1.isEvmStep)(step) || (0, message_trace_1.isPrecompileTrace)(step))
                return;
            if ((0, message_trace_1.isCreateTrace)(step) && step.deployedContract) {
                const address = `0x${step.deployedContract.toString('hex')}`;
                const bytecode = await this.provider.getCode(address);
                parsedTrace.push({
                    op: 'CREATE2',
                    contractHashedBytecode: (0, ethers_1.keccak256)(bytecode),
                    address: address,
                    depth: step.depth
                });
            }
            if ((0, message_trace_1.isCallTrace)(step)) {
                const address = `0x${step.address.toString('hex')}`;
                const bytecode = await this.provider.getCode(address);
                parsedTrace.push({
                    op: 'CALL',
                    contractHashedBytecode: (0, ethers_1.keccak256)(bytecode),
                    address: address,
                    input: step.calldata.toString('hex'),
                    depth: step.depth,
                    returnData: step.returnData.toString('hex')
                });
            }
            for (var i = 0; i < step.steps.length; i++) {
                await stepper(step.steps[i]);
            }
        };
        if (!(0, message_trace_1.isEvmStep)(trace) && !(0, message_trace_1.isPrecompileTrace)(trace)) {
            for (const step of trace.steps) {
                await stepper(step);
            }
        }
        this.traces.push(parsedTrace);
    }
    async resetWorkspace(workspace) {
        if (this.env.config.ethernal.disabled) {
            return;
        }
        const envSet = await this.setLocalEnvironment();
        if (!envSet) {
            return;
        }
        logger(`Resetting workspace "${workspace}"...`);
        try {
            await this.api.resetWorkspace(workspace);
            logger(`Workspace "${workspace}" has been reset!`);
        }
        catch (error) {
            handleError(`Error while resetting workspace "${workspace}"`, error, this.verbose);
        }
    }
    async onData(blockNumber, error) {
        if (error && error.reason) {
            return logger(`Error while receiving data: ${error.reason}`);
        }
        if (this.env.config.ethernal.serverSync) {
            try {
                logger(`Syncing block #${blockNumber}...`);
                await this.api.syncBlock({ number: blockNumber }, true);
            }
            catch (error) {
                handleError(`Couldn't sync block #${blockNumber}`, error, this.verbose);
            }
        }
        else {
            try {
                const block = await this.provider.getBlock(blockNumber, true);
                await this.syncBlock(block);
            }
            catch (error) {
                handleError(`Couldn't sync block #${blockNumber}`, error, this.verbose);
            }
        }
    }
    async setLocalEnvironment() {
        if (this.api.isLoggedIn && this.api.hasWorkspace)
            return true;
        if (!this.api.isLoggedIn) {
            const isLoggedIn = await this.login();
            if (!isLoggedIn)
                return false;
        }
        const isWorkspaceSet = await this.setWorkspace();
        if (!isWorkspaceSet)
            return false;
        return true;
    }
    async syncBlock(block) {
        if (block) {
            const trace = this.traces.shift();
            try {
                await this.api.syncBlock(this.stringifyBns(block), false);
                logger(`Synced block #${block.number}`);
            }
            catch (error) {
                handleError(`Couldn't sync block #${block.number}`, error, this.verbose);
            }
            for (let i = 0; i < block.transactions.length; i++) {
                const hash = block.transactions[i];
                const transaction = await this.provider.getTransaction(hash);
                try {
                    const receipt = await this.provider.getTransactionReceipt(transaction.hash);
                    if (!receipt)
                        logger(`Couldn't get receipt for transaction ${transaction.hash}`);
                    else {
                        const syncedTransaction = Object.assign(Object.assign({}, this.stringifyBns(this.sanitize(transaction))), { provider: null, r: transaction.signature.r, s: transaction.signature.s, transactionIndex: transaction.index || transaction.transactionIndex });
                        // We can't match a trace to a transaction, so we assume blocks with 1 transaction, and sync the trace only for the first one
                        await this.syncTransaction(block, syncedTransaction, this.stringifyBns(this.sanitize(Object.assign(Object.assign({}, receipt), { logs: receipt.logs, provider: null }))), i == 0 ? trace : null);
                    }
                }
                catch (error) {
                    handleError(`Couldn't sync transaction ${transaction.hash}`, error, this.verbose);
                }
            }
        }
    }
    async syncTransaction(block, transaction, transactionReceipt, trace) {
        try {
            await this.api.syncTransaction(block, transaction, transactionReceipt);
            logger(`Synced transaction ${transaction.hash}`);
        }
        catch (error) {
            handleError(`Couldn't sync transaction ${transaction.hash}`, error, this.verbose);
        }
        if (trace && trace.length) {
            try {
                await this.api.syncTrace(transaction.hash, trace);
                logger(`Synced trace for transaction ${transaction.hash}`);
            }
            catch (error) {
                handleError(`Error while syncing trace for transaction ${transaction.hash}`, error, this.verbose);
            }
        }
    }
    async setWorkspace() {
        try {
            const workspace = await this.api.setWorkspace(this.env.config.ethernal.workspace);
            logger(`Using workspace "${workspace.name}"`);
            return true;
        }
        catch (error) {
            handleError(`Error while setting the workspace`, error, this.verbose);
            return false;
        }
    }
    async login() {
        try {
            if (this.env.config.ethernal.apiToken) {
                await this.api.setApiToken(this.env.config.ethernal.apiToken);
                logger(`Authenticated with API token.`);
            }
            else {
                let email, password;
                email = this.env.config.ethernal.email;
                if (!email) {
                    return logger(`Missing email to authenticate. Make sure you've either set ETHERNAL_EMAIL in your environment or they key 'email' in your Ethernal config object & restart the node.`);
                }
                else {
                    password = this.env.config.ethernal.password;
                    if (!password) {
                        return logger(`Missing password to authenticate. Make sure you've either set ETHERNAL_PASSWORD in your environment or they key 'password' in your Ethernal config object & restart the node.`);
                    }
                }
                await this.api.login(email, password);
                logger(`Logged in with ${email}`);
            }
            return true;
        }
        catch (error) {
            handleError(`Login error`, error, this.verbose);
            return false;
        }
    }
    async getFormattedArtifact(targetContract) {
        const fullyQualifiedNames = [];
        (await this.env.artifacts.getAllFullyQualifiedNames()).map(fqn => {
            if (fqn.split(':')[1] == targetContract.name)
                fullyQualifiedNames.push(fqn);
        });
        var defaultBuildInfo = {
            output: {
                contracts: {},
                sources: {}
            }
        };
        let res = {
            name: '',
            address: '',
            abi: {},
            artifact: '',
            dependencies: {}
        };
        for (var i = 0; i < fullyQualifiedNames.length; i++) {
            var buildInfo = await this.env.artifacts.getBuildInfo(fullyQualifiedNames[i]);
            if (!buildInfo) {
                continue;
            }
            var buildInfoContracts = buildInfo.output.contracts;
            var buildInfoOutputSources = buildInfo.output.sources;
            var buildInfoInputSources = buildInfo.input.sources;
            for (var contractFile in buildInfoContracts) {
                for (var contractName in buildInfoContracts[contractFile]) {
                    var artifact = JSON.stringify({
                        contractName: contractName,
                        abi: buildInfoContracts[contractFile][contractName].abi,
                        ast: buildInfoOutputSources[contractFile].ast,
                        source: buildInfoInputSources[contractFile].content
                    });
                    if (contractName == targetContract.name) {
                        res.abi = buildInfoContracts[contractFile][contractName].abi;
                        res.name = contractName;
                        res.artifact = artifact;
                        res.address = targetContract.address;
                    }
                    else {
                        res.dependencies[contractName] = artifact;
                    }
                }
            }
        }
        return res;
    }
    sanitize(obj) {
        var res = {};
        res = Object.fromEntries(Object.entries(obj)
            .filter(([k, v]) => v != null && k != 'provider'));
        return res;
    }
    stringifyBns(obj) {
        var res = {};
        for (const key in obj) {
            if (typeof obj[key] == 'bigint') {
                res[key] = obj[key].toString();
            }
            else {
                res[key] = obj[key];
            }
        }
        return res;
    }
}
exports.Ethernal = Ethernal;
